### PR TITLE
AST-35640 | Create Groups Assignment

### PR DESCRIPTION
**Pull request by tamarleviCm**
### Description

> Adding the new access management group assignment

### References

> https://checkmarx.atlassian.net/browse/AST-35640

### Testing

> added unit & Integration tests
> Manual e2e tests done

### Checklist

- [x] I have added documentation for new/changed functionality in this PR (if applicable).
- [x] I have updated the CLI help for new/changed functionality in this PR (if applicable).
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used

### CODE REVIEW

File: internal/commands/groups.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+func findGroupID(groups []wrappers.Group, name string) wrappers.Group {
+	for i := 0; i < len(groups); i++ {
+		if groups[i].Name == name {
+			return groups[i]
+		}
+	}
+	return wrappers.Group{}
+}
+
+func updateGroupValues(input *[]byte, cmd *cobra.Command, groupsWrapper wrappers.GroupsWrapper) error {
```
Comment (OrShamirCM) [inline internal/commands/groups.go:52]: The FF check is supposed to be in this class right?

File: internal/wrappers/access-management-http.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+			ResourceName: projectName,
+		}
+		params, err := json.Marshal(assignment)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to parse request body")
+		}
+		path := fmt.Sprintf("%s/%s", a.path, createAssignmentPath)
+		resp, err = SendHTTPRequestWithJSONContentType(http.MethodPost, path, bytes.NewBuffer(params), true, a.clientTimeout)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to create groups assignment")
```
Comment (OrShamirCM) [inline internal/wrappers/access-management-http.go:54]: log? or debug log?

File: internal/commands/groups.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+func assignGroupsToProject(projectID string, projectName string, groups []*wrappers.Group,
+	accessManagement wrappers.AccessManagementWrapper) error {
+	if !wrappers.FeatureFlags[accessManagementEnabled] {
+		return nil
+	}
+	log.Println("Creating groups assignment...")
+	groupsAssignedToTheProject, err := accessManagement.GetGroups(projectID)
+	if err != nil {
+		return err
```
Comment (OrShamirCM) [inline internal/commands/groups.go:89]: when error add logs please

File: internal/wrappers/access-management-http.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+		}
+		defer resp.Body.Close()
+	}
+
+	return nil
+}
+
+func (a *AccessManagementHTTPWrapper) GetGroups(projectID string) ([]*Group, error) {
+	log.Println("Getting groups")
+	path := fmt.Sprintf("%s/%s?resource-id=%s&entity-types=group", a.path, entitiesForPath, projectID)
```
Comment (OrShamirCM) [inline internal/wrappers/access-management-http.go:64]: it's resource-type and not entity type, right?

File: internal/wrappers/access-management-http.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	resp, err := SendHTTPRequest(http.MethodGet, path, nil, true, a.clientTimeout)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get groups")
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
```
Comment (OrShamirCM) [inline internal/wrappers/access-management-http.go:70]: do we map and parse the reponse?

File: internal/commands/groups.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	return groupMap, nil
+}
+
+func findGroupID(groups []wrappers.Group, name string) wrappers.Group {
+	for i := 0; i < len(groups); i++ {
+		if groups[i].Name == name {
+			return groups[i]
+		}
+	}
+	return wrappers.Group{}
```
[the remark this answers is not in the document]
Comment (tiagobcx) [reply internal/commands/groups.go:47]: you can use a pointer to wrappers.Group and follow @OrShamirCM  suggestion

File: internal/commands/groups.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	if !wrappers.FeatureFlags[accessManagementEnabled] {
+		return nil
+	}
+	groupsAssignedToTheProject, err := accessManagement.GetGroups(projectID)
+	if err != nil {
+		return err
+	}
+	groupsToAssign := getGroupsToAssign(groups, groupsAssignedToTheProject)
+	if len(groupsToAssign) == 0 {
+		log.Println("No new groups to assign")
```
Comment (tiagobcx) [inline internal/commands/groups.go:84]: Do we want to see this log on the terminal? Maybe more a product question but I would only print it in verbose mode. There is a function printIfVerbose

File: internal/commands/groups.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	var groupsMap = make(map[string]bool)
+	for _, existingGroup := range existingGroups {
+		groupsMap[existingGroup.ID] = true
+	}
+	for _, receivedGroup := range receivedGroups {
+		find := groupsMap[receivedGroup.ID]
+		if !find {
+			groupsToAssign = append(groupsToAssign, receivedGroup)
+		} else {
+			log.Printf("Group [%s | %s] already assigned", receivedGroup.ID, receivedGroup.Name)
```
Comment (tiagobcx) [inline internal/commands/groups.go:106]: Same as above, not sure if we want to see this information unless on debug mode

File: internal/wrappers/access-management-http.go (hunk 2 above)
Comment (tiagobcx) [reply internal/wrappers/access-management-http.go:53]: I think it should be debug or even no print since we will have the error

**Review by tiagobcx — CHANGES_REQUESTED** [review]

File: internal/commands/groups.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	accessManagement wrappers.AccessManagementWrapper) error {
+	if !wrappers.FeatureFlags[accessManagementEnabled] {
+		return nil
+	}
+	log.Println("Creating groups assignment...")
+	groupsAssignedToTheProject, err := accessManagement.GetGroups(projectID)
+	if err != nil {
+		return err
+	}
+	groupsToAssign := getGroupsToAssign(groups, groupsAssignedToTheProject)
```
[the remark this answers is not in the document]
Comment (OrShamirCM) [reply internal/commands/groups.go:82]: I understand, I want the code to be self-explained. 
GetUnassinedGroups maybe ?

File: internal/commands/groups.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+const accessManagementEnabled = "ACCESS_MANAGEMENT_ENABLED" // feature flag
+
+func createGroupsMap(groupsStr string, groupsWrapper wrappers.GroupsWrapper) ([]*wrappers.Group, error) {
+	groups := strings.Split(groupsStr, ",")
+	var groupMap []*wrappers.Group
+	var groupsNotFound []string
+	for _, group := range groups {
+		if len(group) > 0 {
+			groupIds, err := groupsWrapper.Get(group)
```
[the remark this answers is not in the document]
Comment (OrShamirCM) [reply internal/commands/groups.go]: Not sure I understand

**Comment by OrShamirCM** [comment]
Please update the description with more "user (customer) friendly" description (adding the ability for using the new access management for groups association) and check the checkbox

File: internal/commands/project.go [hunk 11]
```diff
@@ -324,7 +269,12 @@ func runCreateProjectCommand(
 				return errors.Wrapf(err, "%s", failedCreatingProj)
 			}
 		}
-
+		if wrappers.FeatureFlags[accessManagementEnabled] {
```
Comment (OrShamirCM) [inline internal/commands/project.go:272]: should it be in groups, no? then it doesn't matter if it called from create project/ update

**Review by OrShamirCM — COMMENTED** [review]
good work

**Review by OrShamirCM — APPROVED** [review]

**Review by tiagobcx — APPROVED** [review]

**Pull request merged**